middle_end/flambda2/types/grammar/more_type_creators.ml

```diff
@@ -352,3 +352,27 @@ let unknown_types_from_arity arity =
   List.map
     (unknown_with_subkind ?alloc_mode:None)
     (Flambda_arity.unarized_components arity)
+
+let shape_of_relation immediates (rel : TG.relation) : _ Or_unknown_or_bottom.t
+    =
+  let module I = Targetint_31_63 in
+  match rel with
+  | Is_int name -> (
+    match I.Set.mem I.zero immediates, I.Set.mem I.one immediates with
+    | true, true -> Unknown
+    | false, false -> Bottom
+    | true, false -> Ok (name, any_block)
+    | false, true -> Ok (name, any_tagged_immediate))
+  | Get_tag name -> (
+    let tags =
+      I.Set.fold
+        (fun tag tags ->
+          match Tag.create_from_targetint tag with
+          | Some tag -> Tag.Set.add tag tags
+          | None -> tags
```

Comment:
Shouldn't we produce `bottom` in that case ? From what I understand, we are trying to reconstruct a shape from information about a `Gat_tag`, and in this case, we have just seen a tag that is outside the bounds of tags, which should not happen except in dead branches ?

Reply:
Returning Bottom for the whole type would be unsound, because you could end up with Bottom here just because you've lost precision elsewhere.
Not adding the tag to the set already treats the corresponding case as unreachable.

Reply:
Indeed, I misunderstood.